src/librustc_typeck/check/mod.rs

```diff
@@ -2990,6 +3031,35 @@ enum TupleArgumentsFlag {
     TupleArguments,
 }
 
+fn expected_types_for_fn_args<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
+                                        call_span: Span,
+                                        expected_ret: Expectation<'tcx>,
+                                        formal_ret: ty::FnOutput<'tcx>,
+                                        formal_args: &[Ty<'tcx>])
+                                        -> Vec<Ty<'tcx>> {
+    // Unify the return type with the expected type early, for more coercions.
+    let expected_args = expected_ret.only_has_type(fcx).and_then(|ret_ty| {
+        if let ty::FnConverging(formal_ret_ty) = formal_ret {
+            fcx.infcx().commit_regions_if_ok(|| {
+                let origin = infer::Misc(call_span);
+                fcx.infcx().sub_types(false, origin, formal_ret_ty, ret_ty).map(|_| {
```

Comment:
can we do `let () = try!(fcx.infcx().sub_types(false, origin, formal_ret_ty, ret_ty));`